File: passman-core/src/buttercup/decrypt.rs
use super::types::*;
use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Key, Nonce,
};
use base64::engine::{general_purpose::STANDARD as BASE64, Engine};
use flate2::read::GzDecoder;
use hmac::digest::KeyInit as HmacKeyInit;
use hmac::{Hmac, Mac};
use pbkdf2::pbkdf2_hmac;
use sha2::Sha256;
use std::io::Read;

pub(super) fn parse_encrypted_components(
    encrypted_text: &str,
) -> Result<EncryptedComponents, ButtercupError> {
    let parts: Vec<&str> = encrypted_text.split('$').collect();
    if parts.len() < 5 {
        return Err(ButtercupError::InvalidFormat);
    }

    let rounds = parts[4]
        .parse::<u32>()
        .map_err(|_| ButtercupError::InvalidRounds(parts[4].to_string()))?;

    let method = if parts.len() == 5 {
        DEFAULT_ALGORITHM.to_string()
    } else {
        parts[5].to_string().to_lowercase()
    };

    Ok(EncryptedComponents {
        content: parts[0].to_string(),
        iv: parts[1].to_string(),
        salt: parts[2].to_string(),
        auth: parts[3].to_string(),
        rounds,
        method,
    })
}

pub(super) fn decrypt_components(
    components: &EncryptedComponents,
    password: &str,
) -> Result<String, ButtercupError> {
    match components.method.as_str() {
        "cbc" => decrypt_cbc(components, password),
        "gcm" => decrypt_gcm(components, password),
        other => Err(ButtercupError::UnsupportedAlgorithm(other.to_string())),
    }
}

type DecodedComponents = (Vec<u8>, Vec<u8>, Vec<u8>);

fn decode_common_components(
    components: &EncryptedComponents,
) -> Result<DecodedComponents, ButtercupError> {
    let content_bytes = BASE64.decode(&components.content)?;
    let iv_bytes = hex::decode(&components.iv)?;
    let auth_bytes = hex::decode(&components.auth)?;
    Ok((content_bytes, iv_bytes, auth_bytes))
}

fn decrypt_cbc(components: &EncryptedComponents, password: &str) -> Result<String, ButtercupError> {
    let salt_bytes = components.salt.as_bytes();
    let derived = derive_key(
        password,
        salt_bytes,
        components.rounds,
        PASSWORD_KEY_SIZE + HMAC_KEY_SIZE,
    );
    let (key, hmac_key) = derived.split_at(PASSWORD_KEY_SIZE);

    let (content_bytes, iv_bytes, auth_bytes) = decode_common_components(components)?;

    let mut mac = <Hmac<Sha256> as HmacKeyInit>::new_from_slice(hmac_key)
        .map_err(|e| ButtercupError::DecryptionFailed(e.to_string()))?;
    mac.update(components.content.as_bytes());
    mac.update(components.iv.as_bytes());
    mac.update(components.salt.as_bytes());
    let result = mac.finalize();
    let expected_hmac = result.into_bytes();
    if expected_hmac.as_slice() != auth_bytes.as_slice() {
        return Err(ButtercupError::AuthenticationFailed);
    }

    let cipher = cbc::Decryptor::<aes::Aes256>::new(key.into(), iv_bytes.as_slice().into());
    let mut buffer = content_bytes.to_vec();
    let decrypted = cipher
        .decrypt_padded_mut::<Pkcs7>(&mut buffer)
        .map_err(|e| ButtercupError::DecryptionFailed(e.to_string()))?;

    String::from_utf8(decrypted.to_vec()).map_err(ButtercupError::Utf8)
}

fn decrypt_gcm(components: &EncryptedComponents, password: &str) -> Result<String, ButtercupError> {
    let salt_bytes = components.salt.as_bytes();
    let key = derive_key(password, salt_bytes, components.rounds, PASSWORD_KEY_SIZE);

    let (content_bytes, iv_bytes, auth_tag_bytes) = decode_common_components(components)?;

    let cipher_key = Key::<Aes256Gcm>::from_slice(&key);
    let cipher = Aes256Gcm::new(cipher_key);
    let nonce = Nonce::from_slice(iv_bytes.as_slice());

    let mut full_ciphertext = content_bytes;
    full_ciphertext.extend_from_slice(&auth_tag_bytes);

    let aad = format!("{}{}", components.iv, components.salt).into_bytes();

    let decrypted = cipher
        .decrypt(
            nonce,
            aes_gcm::aead::Payload {
                msg: &full_ciphertext,
                aad: &aad,
            },
        )
        .map_err(|e| ButtercupError::DecryptionFailed(e.to_string()))?;

    String::from_utf8(decrypted).map_err(ButtercupError::Utf8)
}

fn derive_key(password: &str, salt: &[u8], rounds: u32, length: usize) -> Vec<u8> {
    let mut output = vec![0u8; length];
    pbkdf2_hmac::<Sha256>(password.as_bytes(), salt, rounds, &mut output);
    output
}

pub(super) fn decompress(compressed: &str) -> Result<String, ButtercupError> {
    let compressed_bytes = BASE64.decode(compressed)?;
    let mut decoder = GzDecoder::new(compressed_bytes.as_slice());
    let mut output = String::new();
    decoder
        .read_to_string(&mut output)
        .map_err(|e| ButtercupError::DecompressionFailed(e.to_string()))?;
    Ok(output)
}

File: pkg/buttercup/decrypt.go
package buttercup

import (
	"bytes"
	"compress/gzip"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"io"
	"strconv"
	"strings"

	"golang.org/x/crypto/pbkdf2"
)

func ParseEncryptedComponents(encryptedText string) (*EncryptedComponents, error) {
	parts := strings.Split(encryptedText, "$")
	if len(parts) < 5 {
		return nil, &ButtercupError{Msg: "invalid encrypted format"}
	}

	rounds, err := strconv.ParseUint(parts[4], 10, 32)
	if err != nil {
		return nil, &ButtercupError{Msg: fmt.Sprintf("invalid rounds: %s", parts[4])}
	}

	method := DefaultAlgorithm
	if len(parts) >= 6 {
		method = strings.ToLower(parts[5])
	}

	return &EncryptedComponents{
		Content: parts[0],
		IV:      parts[1],
		Salt:    parts[2],
		Auth:    parts[3],
		Rounds:  uint32(rounds),
		Method:  method,
	}, nil
}

func DecryptComponents(components *EncryptedComponents, password string) (string, error) {
	switch components.Method {
	case "cbc":
		return decryptCBC(components, password)
	case "gcm":
		return decryptGCM(components, password)
	default:
		return "", &ButtercupError{Msg: fmt.Sprintf("unsupported algorithm: %s", components.Method)}
	}
}

func decodeCommonComponents(c *EncryptedComponents) ([]byte, []byte, []byte, error) {
	contentBytes, err := base64.StdEncoding.DecodeString(c.Content)
	if err != nil {
		return nil, nil, nil, err
	}
	ivBytes, err := hex.DecodeString(c.IV)
	if err != nil {
		return nil, nil, nil, err
	}
	authBytes, err := hex.DecodeString(c.Auth)
	if err != nil {
		return nil, nil, nil, err
	}
	return contentBytes, ivBytes, authBytes, nil
}

func deriveKey(password string, salt []byte, rounds uint32, length int) []byte {
	return pbkdf2.Key([]byte(password), salt, int(rounds), length, sha256.New)
}

func decryptCBC(c *EncryptedComponents, password string) (string, error) {
	saltBytes := []byte(c.Salt)
	derived := deriveKey(password, saltBytes, c.Rounds, PasswordKeySize+HMACKeySize)
	key := derived[:PasswordKeySize]
	hmacKey := derived[PasswordKeySize:]

	contentBytes, ivBytes, authBytes, err := decodeCommonComponents(c)
	if err != nil {
		return "", err
	}

	mac := hmac.New(sha256.New, hmacKey)
	mac.Write([]byte(c.Content))
	mac.Write([]byte(c.IV))
	mac.Write([]byte(c.Salt))
	expectedHMAC := mac.Sum(nil)
	if !hmac.Equal(expectedHMAC, authBytes) {
		return "", &ButtercupError{Msg: "authentication failed"}
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}
	if len(ivBytes) != aes.BlockSize {
		return "", &ButtercupError{Msg: "invalid IV length"}
	}
	mode := cipher.NewCBCDecrypter(block, ivBytes)

	plaintext := make([]byte, len(contentBytes))
	mode.CryptBlocks(plaintext, contentBytes)

	plaintext, err = pkcs7Unpad(plaintext, aes.BlockSize)
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}

	return string(plaintext), nil
}

func decryptGCM(c *EncryptedComponents, password string) (string, error) {
	saltBytes := []byte(c.Salt)
	key := deriveKey(password, saltBytes, c.Rounds, PasswordKeySize)

	contentBytes, ivBytes, authTagBytes, err := decodeCommonComponents(c)
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}

	fullCiphertext := append(contentBytes, authTagBytes...)
	aad := []byte(c.IV + c.Salt)

	plaintext, err := gcm.Open(nil, ivBytes, fullCiphertext, aad)
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}

	return string(plaintext), nil
}

func Decompress(compressed string) (string, error) {
	compressedBytes, err := base64.StdEncoding.DecodeString(compressed)
	if err != nil {
		return "", err
	}
	r, err := gzip.NewReader(bytes.NewReader(compressedBytes))
	if err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}
	defer r.Close()
	var buf bytes.Buffer
	if _, err := io.Copy(&buf, r); err != nil {
		return "", &ButtercupError{Msg: err.Error()}
	}
	return buf.String(), nil
}

func pkcs7Unpad(data []byte, blockSize int) ([]byte, error) {
	if len(data) == 0 {
		return nil, fmt.Errorf("invalid padding")
	}
	padding := int(data[len(data)-1])
	if padding < 1 || padding > blockSize {
		return nil, fmt.Errorf("invalid padding")
	}
	if len(data) < padding {
		return nil, fmt.Errorf("invalid padding")
	}
	return data[:len(data)-padding], nil
}
